### PR TITLE
Prefer concatenate to h/vstack in simple cases.

### PR DESCRIPTION
**Pull request by anntzer**
concatenate is (much) faster than hstack, while being essentially
equivalent in the 1D case; when working on 2D arrays it can replace
vstack too.

Timings:
```
In [1]: u = np.arange(5)

In [2]: %timeit np.concatenate([u, u])
842 ns ± 8.16 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [3]: %timeit np.hstack([u, u])
2.43 µs ± 7.43 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [4]: u = np.arange(1000)

In [5]: %timeit np.concatenate([u, u])
1.2 µs ± 2.67 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [6]: %timeit np.hstack([u, u])
2.9 µs ± 10.1 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [7]: u = np.arange(10).reshape((5, 2))

In [8]: %timeit np.concatenate([u, u])
934 ns ± 19.9 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [9]: %timeit np.vstack([u, u])
2.54 µs ± 7.55 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [10]: u = np.arange(2000).reshape((1000, 2))

In [11]: %timeit np.concatenate([u, u])
1.61 µs ± 2.64 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [12]: %timeit np.vstack([u, u])
3.37 µs ± 17.1 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

## PR Checklist

- [ ] Has pytest style unit tests (and `pytest` passes).
- [ ] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (run `flake8` on changed files to check).
- [ ] New features are documented, with examples if plot related.
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [ ] Conforms to Matplotlib style conventions (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).
- [ ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [ ] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).

### CODE REVIEW

File: lib/matplotlib/patches.py [hunk 1]
```diff
@@ -994,14 +994,14 @@ def _update_path(self):
...
             elif self._baseline.ndim == 0:  # single baseline value
-                y = np.hstack((self._baseline, y, self._baseline))
+                y = np.concatenate([[self._baseline], y, [self._baseline]])
             elif self._baseline.ndim == 1:  # baseline array
                 base = np.repeat(self._baseline[idx0:idx1], 2)[::-1]
                 x = np.concatenate([x, x[::-1]])
-                y = np.concatenate([np.hstack((base[-1], y, base[0],
-                                               base[0], base, base[-1]))])
+                y = np.concatenate([[base[-1]], y, [base[0]],
+                                    [base[0]], base, [base[-1]]])
```
Comment (QuLogic) [inline lib/matplotlib/patches.py:1004]: And?
```suggestion
                y = np.concatenate([base[-1:], y, base[:1],
                                    base[:1], base, base[-1:]])
```

**Review by QuLogic — APPROVED** [review]

File: lib/matplotlib/patches.py [hunk 2]
```diff
@@ -994,14 +994,14 @@ def _update_path(self):
             x = np.repeat(self._edges[idx0:idx1+1], 2)
             y = np.repeat(self._values[idx0:idx1], 2)
             if self._baseline is None:
-                y = np.hstack((y[0], y, y[-1]))
+                y = np.concatenate([[y[0]], y, [y[-1]]])
```
[the remark this answers is not in the document]
Comment (anntzer) [reply lib/matplotlib/patches.py:997]: yes

File: lib/matplotlib/patches.py (hunk 1 above)
Comment (anntzer) [reply lib/matplotlib/patches.py:1004]: yes

File: lib/matplotlib/markers.py [hunk 3]
```diff
@@ -641,10 +641,10 @@ def _set_hexagon1(self):
 
             # not drawing inside lines
             x = np.abs(np.cos(5 * np.pi / 6.))
-            top = Path(np.vstack(([-x, 0], verts[(1, 0, 5), :], [x, 0])))
-            bottom = Path(np.vstack(([-x, 0], verts[2:5, :], [x, 0])))
-            left = Path(verts[(0, 1, 2, 3), :])
-            right = Path(verts[(0, 5, 4, 3), :])
+            top = Path(np.concatenate([[(-x, 0)], verts[[1, 0, 5]], [(x, 0)]]))
+            bottom = Path(np.concatenate([[(-x, 0)], verts[2:5], [(x, 0)]]))
+            left = Path(verts[[0, 1, 2, 3]])
```
Comment (timhoffm) [inline lib/matplotlib/markers.py:646]:
```suggestion
            left = Path(verts[:4])
```

File: lib/matplotlib/markers.py [hunk 4]
```diff
@@ -675,11 +675,12 @@ def _set_hexagon2(self):
 
             # not drawing inside lines
             x, y = np.sqrt(3) / 4, 3 / 4.
-            top = Path(verts[(1, 0, 5, 4, 1), :])
-            bottom = Path(verts[(1, 2, 3, 4), :])
-            left = Path(np.vstack(([x, y], verts[(0, 1, 2), :],
-                                   [-x, -y], [x, y])))
-            right = Path(np.vstack(([x, y], verts[(5, 4, 3), :], [-x, -y])))
+            top = Path(verts[[1, 0, 5, 4, 1]])
+            bottom = Path(verts[[1, 2, 3, 4]])
```
Comment (timhoffm) [inline lib/matplotlib/markers.py:679]:
```suggestion
            bottom = Path(verts[1:5])
```

File: lib/matplotlib/markers.py [hunk 5]
```diff
@@ -675,11 +675,12 @@ def _set_hexagon2(self):
...
+            left = Path(np.concatenate([
+                [(x, y)], verts[[0, 1, 2]], [(-x, -y), (x, y)]]))
```
Comment (timhoffm) [inline lib/matplotlib/markers.py:681]:
```suggestion
                [(x, y)], verts[:3], [(-x, -y), (x, y)]]))
```

File: lib/matplotlib/markers.py [hunk 6]
```diff
@@ -675,11 +675,12 @@ def _set_hexagon2(self):
...
+            right = Path(np.concatenate([
+                [(x, y)], verts[[5, 4, 3]], [(-x, -y)]]))
```
Comment (timhoffm) [inline lib/matplotlib/markers.py:683]:
```suggestion
                [(x, y)], verts[5:2:-1], [(-x, -y)]]))
```

File: lib/matplotlib/markers.py (hunk 3 above)
Comment (anntzer) [reply lib/matplotlib/markers.py]: yes to all

**Review by timhoffm — APPROVED** [review]

**Pull request merged**